pkg/webhook/instance_admission.go

```diff
@@ -29,99 +34,148 @@ func (ia *InstanceAdmission) Handle(ctx context.Context, req admission.Request)
-		// and if it's an upgrade, we need the new one anyway
-		ov, err := instance.GetOperatorVersion(new, ia.client)
-		if err != nil {
-			log.Printf("InstanceAdmission: Error getting operatorVersion %s for instance %s/%s: %v", new.Spec.OperatorVersion.Name, new.Namespace, new.Name, err)
-			admission.Errored(http.StatusInternalServerError, err)
-		}
+	// if there is a 'cleanup' plan present in the OV, we add a finalizer to the instance
+	if kudov1beta1.CleanupPlanExists(ov) {
+		new.TryAddFinalizer()
+	}
 
-		triggered, err := admitUpdate(old, new, ov)
-		if err != nil {
-			return admission.Denied(err.Error())
-		}
+	// add an instance snapshot *BEFORE* setting new Spec.PlanExecution. this way the controller will recognize the plan as newly scheduled
+	if err = new.AnnotateSnapshot(); err != nil {
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failed to create an Instance snapshot %s/%s: %v", new.Namespace, new.Name, err))
+	}
 
-		// Populate Instance.PlanExecution with the plan triggered by param change
-		if triggered != nil {
-			new.Spec.PlanExecution.PlanName = *triggered
-		}
+	// schedule 'deploy' plan for execution (and fail if it doesn't exist)
+	if !kudov1beta1.PlanExists(kudov1beta1.DeployPlanName, ov) {
+		return admission.Denied(fmt.Sprintf("failed to create an Instance %s/%s: couldn't find '%s' plan in the operatorVersion", new.Namespace, new.Name, kudov1beta1.DeployPlanName))
+	}
+	new.Spec.PlanExecution.PlanName = kudov1beta1.DeployPlanName
+	new.Spec.PlanExecution.UID = uuid.NewUUID()
 
-		return admission.Allowed("")
+	marshaled, err := json.Marshal(new)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
 
-	default:
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaled)
+}
+
+func handleUpdate(ia *InstanceAdmission, req admission.Request) admission.Response {
+	old, new := &kudov1beta1.Instance{}, &kudov1beta1.Instance{}
+
+	// req.Object contains the updated object
+	if err := ia.decoder.DecodeRaw(req.Object, new); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// req.OldObject is populated for DELETE and UPDATE requests
+	if err := ia.decoder.DecodeRaw(req.OldObject, old); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// fetch new OperatorVersion: we always fetch the new one, since if it's an update it's the same as the old one
+	// and if it's an upgrade, we need the new one anyway
+	ov, err := instance.GetOperatorVersion(new, ia.client)
+	if err != nil {
+		log.Printf("InstanceAdmission: Error getting operatorVersion %s for instance %s/%s: %v", new.Spec.OperatorVersion.Name, new.Namespace, new.Name, err)
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	// we explicitly ignore Metadata updates
+	if reflect.DeepEqual(old.Spec, new.Spec) && reflect.DeepEqual(old.Status, new.Status) {
```

Comment:
smart :)

Reply:
🙇

Reply:
It does, while the instance `generation` changes. Admission webhook uses this to reset `PlanExecution` when a plan is terminal (see L265 in this file)